src/mixins/object_interactivity.mixin.ts

```diff
@@ -1,394 +1,497 @@
+    }
+    return false;
+  }
+
+  /**
+   * Calculates the coordinates of the center of each control plus the corners of the control itself
+   * This basically just delegates to each control positionHandler
+   * WARNING: changing what is passed to positionHandler is a breaking change, since position handler
+   * is a public api and should be done just if extremely necessary
+   * @return {Record<string, TOCoord>}
+   */
+  calcOCoords(): Record<string, TOCoord> {
+    const vpt = this.getViewportTransform(),
+      center = this.getCenterPoint(),
+      tMatrix = [1, 0, 0, 1, center.x, center.y] as TMat2D,
+      rMatrix = calcRotateMatrix({
+        angle: this.getTotalAngle() - (!!this.group && this.flipX ? 180 : 0),
+      }),
+      positionMatrix = multiplyTransformMatrices(tMatrix, rMatrix),
+      startMatrix = multiplyTransformMatrices(vpt, positionMatrix),
+      finalMatrix = multiplyTransformMatrices(startMatrix, [
+        1 / vpt[0],
+        0,
+        0,
+        1 / vpt[3],
+        0,
+        0,
+      ]),
+      transformOptions = this.group
+        ? qrDecompose(this.calcTransformMatrix())
+        : undefined,
+      dim = this._calculateCurrentDimensions(transformOptions),
+      coords: Record<string, TOCoord> = {};
+
+    this.forEachControl((control: any, key: string, fabricObject: InteractiveFabricObject) => {
+      coords[key] = control.positionHandler(dim, finalMatrix, fabricObject);
+    });
+
+    // debug code
+    /*
+      const canvas = this.canvas;
+      setTimeout(function () {
+      if (!canvas) return;
+        canvas.contextTop.clearRect(0, 0, 700, 700);
+        canvas.contextTop.fillStyle = 'green';
+        Object.keys(coords).forEach(function(key) {
+          const control = coords[key];
+          canvas.contextTop.fillRect(control.x, control.y, 3, 3);
         });
-        ctx.restore();
-
-        return this;
-      },
-
-      /**
-       * Returns true if the specified control is visible, false otherwise.
-       * @param {String} controlKey The key of the control. Possible values are 'tl', 'tr', 'br', 'bl', 'ml', 'mt', 'mr', 'mb', 'mtr'.
-       * @returns {Boolean} true if the specified control is visible, false otherwise
-       */
-      isControlVisible: function (controlKey) {
-        return (
-          this.controls[controlKey] &&
-          this.controls[controlKey].getVisibility(this, controlKey)
+      } 50);
+    */
+    return coords;
+  }
+
+  /**
+   * Sets corner and controls position coordinates based on current angle, width and height, left and top.
+   * oCoords are used to find the corners
+   * aCoords are used to quickly find an object on the canvas
+   * lineCoords are used to quickly find object during pointer events.
+   * See {@link https://github.com/fabricjs/fabric.js/wiki/When-to-call-setCoords} and {@link http://fabricjs.com/fabric-gotchas}
+   * @return {void}
+   */
+  setCoords(): void {
+    if (this.callSuper) {
```

Comment:
please explain why this is needed
transient until everything is migrated?
I don't understand why needed since everything funnels through FabricObject

Reply:
fabricCreate.class is assembling a new class from InteractiveObject and creating a function with prototype.
At that point `super.setCoords()` does not work anymore because the method gets called in a context that is not an es6 class. checking if callSuper exists is done to understand if we are using a createClass construct.